pkg/cacheutil/memcached_client.go

```diff
@@ -0,0 +1,382 @@
+type MemcachedClientConfig struct {
+	// Addrs specifies the list of memcached addresses. The addresses get
+	// resolved with the DNS provider.
+	Addrs []string
+
+	// Timeout specifies the socket read/write timeout.
+	Timeout time.Duration
+
+	// MaxIdleConnections specifies the maximum number of idle connections that
+	// will be maintained per address. For better performances, this should be
+	// set to a number higher than your peak parallel requests.
+	MaxIdleConnections int
+
+	// MaxAsyncConcurrency specifies the maximum number of concurrent asynchronous
+	// operations can occur.
+	MaxAsyncConcurrency int
+
+	// MaxAsyncBufferSize specifies the maximum number of enqueued asynchronous
+	// operations allowed.
+	MaxAsyncBufferSize int
+
+	// MaxGetMultiBatchConcurrency specifies the maximum number of concurrent batch
+	// executions by GetMulti().
+	// TODO(pracucci) Should this be a global (per-client) limit or a per-single MultiGet()
+	//                limit? The latter would allow us to avoid a single very large MultiGet()
+	//                will slow down other requests.
+	MaxGetMultiBatchConcurrency int
+
+	// MaxGetMultiBatchSize specified the maximum number of keys a single underlying
+	// GetMulti() should run. If more keys are specified, internally keys are splitted
+	// into multiple batches and fetched concurrently up to MaxGetMultiBatchConcurrency
+	// parallelism.
+	MaxGetMultiBatchSize int
+
+	// DNSProviderUpdateInterval specifies the DNS discovery update interval.
+	DNSProviderUpdateInterval time.Duration
+}
+
+func (c *MemcachedClientConfig) applyDefaults() {
+	if c.Timeout == 0 {
+		c.Timeout = defaultTimeout
+	}
+
+	if c.MaxIdleConnections == 0 {
+		c.MaxIdleConnections = defaultMaxIdleConnections
+	}
+
+	if c.MaxAsyncConcurrency == 0 {
+		c.MaxAsyncConcurrency = defaultMaxAsyncConcurrency
+	}
+
+	if c.MaxAsyncBufferSize == 0 {
+		c.MaxAsyncBufferSize = defaultMaxAsyncBufferSize
+	}
+
+	if c.DNSProviderUpdateInterval == 0 {
+		c.DNSProviderUpdateInterval = defaultDNSProviderUpdateInterval
+	}
+
+	if c.MaxGetMultiBatchConcurrency == 0 {
+		c.MaxGetMultiBatchConcurrency = defaultMaxGetMultiBatchConcurrency
+	}
+
+	if c.MaxGetMultiBatchSize == 0 {
+		c.MaxGetMultiBatchSize = defaultMaxGetMultiBatchSize
+	}
+}
+
+func (c *MemcachedClientConfig) validate() error {
+	if len(c.Addrs) == 0 {
+		return errMemcachedConfigNoAddrs
+	}
+
+	return nil
+}
+
+type memcachedClient struct {
+	logger   log.Logger
+	config   MemcachedClientConfig
+	client   memcachedClientBackend
+	selector *MemcachedJumpHashSelector
+
+	// DNS provider used to keep the memcached servers list updated.
+	provider *dns.Provider
+
+	// Channel used to notify internal goroutines when they should quit.
+	stop chan struct{}
+
+	// Channel used to enqueue async operations.
+	asyncQueue   chan func()
+	asyncWorkers sync.WaitGroup
+
+	// Channel used to enqueue get multi operations.
+	getMultiQueue   chan *memcachedGetMultiBatch
+	getMultiWorkers sync.WaitGroup
```

Comment:
why not use a single wg for all spawned bg routines?

Reply:
Definitely. Fixed.